### PR TITLE
kustomize lookup plugin

### PR DESCRIPTION
**Pull request by abikouo**
##### SUMMARY

new lookup plugin to support kustomize feature for kubernetes

##### ISSUE TYPE

- Feature Pull Request

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/lookup/kustomize.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    short_description: Build a set of kubernetes resources using a 'kustomization.yaml' file.
+
+    author:
+      - Aubin Bikouo <@abikouo>
+    notes:
+      - If both kustomize and kubectl are part of the PATH, kustomize will be used by the plugin.
+    description:
+      - Uses the kustomize or the kubectl tool.
+      - Return the result of kustomize build or kubectl kustomize
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:17]:
```suggestion
      - Return the result of C(kustomize build) or C(kubectl kustomize).
```

File: plugins/lookup/kustomize.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    options:
+      dir:
+        description:
+        - The dir argument must be a path to a directory containing 'kustomization.yaml',
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:21]:
```suggestion
        - The directory path containing 'kustomization.yaml',
```

File: plugins/lookup/kustomize.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+          or a git repository URL with a path suffix specifying same with respect to the repository root.
+        - If dir is omitted, '.' is assumed.
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:23]:
```suggestion
        - If omitted, '.' is assumed.
```

File: plugins/lookup/kustomize.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        default: "."
+      binary_path:
+        description:
+        - The path of a kustomize or kubectl binary to use.
+      opt_dirs:
+        description:
+        - optional list of directories to search for executable in addition to PATH.
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:30]:
```suggestion
        - An optional list of directories to search for the executable in addition to PATH.
```

File: plugins/lookup/kustomize.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    requirements:
+      - "python >= 3.6"
+'''
+
+EXAMPLES = """
+- name: Run lookup using kustomize
+  set_fact:
+    resources: "{{ lookup('kustomize', binary_path='/path/to/kustomize') }}"
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:39]:
```suggestion
    resources: "{{ lookup('kubernetes.core.kustomize', binary_path='/path/to/kustomize') }}"
```

File: plugins/lookup/kustomize.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+- name: Run lookup using kubectl kustomize
+  set_fact:
+    resources: "{{ lookup('kustomize', binary_path='/path/to/kubectl') }}"
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:43]:
```suggestion
    resources: "{{ lookup('kubernetes.core.kustomize', binary_path='/path/to/kubectl') }}"
```

File: plugins/lookup/kustomize.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+- name: Create kubernetes resources for lookup output
+  k8s:
+    definition: "{{ item }}"
+  with_items:
+    "{{ lookup('kustomize', dir='/path/to/kustomization') }}"
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:49]:
```suggestion
    "{{ lookup('kubernetes.core.kustomize', dir='/path/to/kustomization') }}"
```

File: plugins/lookup/kustomize.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    try:
+        opt_arg = {}
+        if opt_dirs is not None:
+            if not isinstance(opt_dirs, list):
+                opt_dirs = [opt_dirs]
+            opt_arg['opt_dirs'] = opt_dirs
+        bin_path = get_bin_path(name, **opt_arg)
+        return bin_path
+    except ValueError:
+        return None
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:100]:
```suggestion
    opt_arg = {}
    try:
        if opt_dirs is not None:
            if not isinstance(opt_dirs, list):
                opt_dirs = [opt_dirs]
            opt_arg['opt_dirs'] = opt_dirs
        return get_bin_path(name, **opt_arg)
    except ValueError:
        return None
```

File: plugins/lookup/kustomize.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        try:
+            executable_path = binary_path
+            if executable_path is None:
+                executable_path = get_binary_from_path(name="kustomize", opt_dirs=opt_dirs)
+                if executable_path is None:
+                    executable_path = get_binary_from_path(name="kubectl", opt_dirs=opt_dirs)
+
+                # validate that at least one tool was found
+                if executable_path is None:
+                    raise AnsibleLookupError("Failed to find required executable 'kubectl' and 'kustomize' in paths")
```
Comment (Akasurde) [inline plugins/lookup/kustomize.py:120]: I think we can move this out of the try-except block.

**Review by Akasurde — DISMISSED** [review]

File: plugins/lookup/kustomize.py (hunk 9 above)
Comment (abikouo) [reply plugins/lookup/kustomize.py:120]: yes agree. Thanks

**Comment by abikouo** [comment]
recheck

File: plugins/lookup/kustomize.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            if executable_path is None:
+                raise AnsibleLookupError("Failed to find required executable 'kubectl' and 'kustomize' in paths")
+
+        # check input directory
+        kustomization_dir = dir
+        if not os.path.isdir(kustomization_dir):
+            raise AnsibleLookupError("dir parameter {0} is not a valid directory.".format(kustomization_dir))
+        kustomization_file = os.path.join(kustomization_dir, "kustomization.yaml")
+        if not os.path.isfile(kustomization_file):
+            raise AnsibleLookupError("missing 'kustomization.yaml' file from input directory '{0}'".format(kustomization_dir))
```
Comment (gravesm) [inline plugins/lookup/kustomize.py:126]: All of this should be removed. We are just duplicating checks that kustomize is already making. This code block is also more restrictive than kustomize. The kustomization file can additionally be named `kustomization.yml` and `Kustomization`.

File: plugins/lookup/kustomize.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+- name: Run lookup using kubectl kustomize
+  set_fact:
+    resources: "{{ lookup('kubernetes.core.kustomize', binary_path='/path/to/kubectl') }}"
+
+- name: Create kubernetes resources for lookup output
+  k8s:
+    definition: "{{ item }}"
+  with_items:
+    "{{ lookup('kubernetes.core.kustomize', dir='/path/to/kustomization') }}"
```
Comment (gravesm) [inline plugins/lookup/kustomize.py:49]: Get rid of the loop here and just apply the lookup directly to the definition. Even with multiple resources it's still only one loop item, and even if it were multiple items it's faster to just use a single definition with multiple resources.

File: plugins/lookup/kustomize.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
+#
+#  Copyright 2021 Red Hat | Ansible
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+DOCUMENTATION = '''
```
Comment (gravesm) [inline plugins/lookup/kustomize.py:6]: We need a version added here.

**Review by gravesm — CHANGES_REQUESTED** [review]

File: plugins/lookup/kustomize.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    lookup: kustomize
+
+    short_description: Build a set of kubernetes resources using a 'kustomization.yaml' file.
+
+    version_added: "1.0.0"
```
Comment (gravesm) [inline plugins/lookup/kustomize.py:11]:
```suggestion
    version_added: "2.2.0"
```
:rewind: :hourglass_flowing_sand:

Comment (abikouo) [reply plugins/lookup/kustomize.py:11]: bad copy paste :-1:

**Review by gravesm — APPROVED** [review]

**Comment by abikouo** [comment]
recheck

**Comment by abikouo** [comment]
recheck

**Comment by gravesm** [comment]
recheck

**Comment by septatrix** [comment]
I noticed that this does not use the a roles `files/` directory be default. Is that intended or just an oversight?

**Pull request merged**